common/remote/sockfile.cpp

```diff
@@ -3811,17 +3832,243 @@ class CClientStatsTable : public OwningStringSuperHashTableOf<CClientStats>
     }
 };
 
+interface IRemoteActivity : extends IInterface
+{
+    virtual const void *nextRow(size32_t &sz) = 0;
+    virtual unsigned __int64 queryProcessed() const = 0;
+    virtual IOutputMetaData *queryOutputMeta() const = 0;
+    virtual StringBuffer &getInfoStr(StringBuffer &out) const = 0;
+    virtual void serializeCursor(MemoryBuffer &tgt) const = 0;
+    virtual void restoreCursor(MemoryBuffer &src) = 0;
+};
+
 enum OpenFileFlag { of_null=0x0, of_key=0x01 };
 struct OpenFileInfo
 {
     OpenFileInfo() { }
     OpenFileInfo(int _handle, IFileIO *_fileIO, StringAttrItem *_filename) : handle(_handle), fileIO(_fileIO), filename(_filename) { }
+    OpenFileInfo(int _handle, IRemoteActivity *_activity, StringAttrItem *_filename) : handle(_handle), activity(_activity), filename(_filename) { }
     Linked<IFileIO> fileIO;
+    Linked<IRemoteActivity> activity;
     Linked<StringAttrItem> filename; // for debug
     int handle = 0;
     unsigned flags = 0;
 };
 
+
+
+static IOutputMetaData *getTypeInfoOutputMetaData(IPropertyTree &actNode, const char *typePropName)
+{
+    IPropertyTree *inputJson = actNode.queryPropTree(typePropName);
+    if (inputJson)
+        return createTypeInfoOutputMetaData(*inputJson);
+    else
+    {
+        StringBuffer binTypePropName(typePropName);
+        MemoryBuffer mb;
+        actNode.getPropBin(binTypePropName.append("Bin").str(), mb);
+        return createTypeInfoOutputMetaData(mb);
+    }
+}
+
+class CRemoteDiskReadActivity : public CSimpleInterfaceOf<IRemoteActivity>
+{
+    StringAttr fileName;
+    Linked<IHThorDiskReadArg> helper;
+    MemoryBuffer resultBuffer;
+    MemoryBufferBuilder *outBuilder = nullptr;
+    unsigned __int64 limit = 0;
+    unsigned __int64 chooseN = 0;
+    IArrayOf<IKeySegmentMonitor> segMonitors;
+    Owned<ISourceRowPrefetcher> prefetcher;
+    CThorContiguousRowBuffer prefetchBuffer;
+    bool opened = false;
+    bool eofseen = false;
+    bool needTransform = true;
+    unsigned __int64 processed = 0;
+    Owned<ISerialStream> inputStream;
+    Owned<IFileIO> iFileIO;
+    unsigned __int64 startPos = 0;
+    Linked<IOutputMetaData> outMeta;
+    bool cursorDirty = false;
+
+    void checkOpen()
+    {
+        if (opened)
+        {
+            if (!cursorDirty)
+                return;
+            if (prefetchBuffer.tell() != startPos)
```

Comment:
this also needs to reset eofseen.  That probably requires the result from canMatchAny() to be stored separately